### ISSUE TITLE
OAuth register endpoint should have CORS enabled

### ISSUE THREAD

**Issue by ptbrowne** [cozy/cozy-stack#1610]
If I try to instantiate an OAuth client from a webpage, it fails because there are no "Access-Control-Allow-Origin" served from the `register` endpoint.

I think this should be relaxed as the `register` endpoint should be accessible from any webpage. 

```
Access to fetch at 'https://recette.cozy.works/auth/register' from origin 'http://localhost:1234' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
_fails.js:8 Uncaught (in promise) TypeError: Failed to fetch
```

**Comment by nono** [comment]
> I think this should be relaxed as the register endpoint should be accessible from any webpage.

Why? Do you think of a use-case for this?

I'm really not cleared about if it's a good idea or not to do that, and what are the security implications of that. I've tried a quick search on Internet but didn't find anything about CORS and the register endpoint.

**Comment by ptbrowne** [comment]
It is also good to test the OAuth part of our applications with no mobile apps.

**Comment by y-lohse** [comment]
👍 on this issue — and possibly disabling CORS on all routes.

Aside from using the stack on pages that are not served by a cozy (which would be useful for documentation examples and code sharing / collaboration, for example via codesandbox.io), this would also help our transition on mobile to the new WebView layer, which does implement CORS (the previous cordova version did not).

CORS are only active on a standard browser anyway and can already be bypassed. The real protection comes from the tokens embeded in the requests — except on the routes related to login/registrations of course. From what I know, it's pretty standard practice to enable CORS on API servers.

**Comment by nono** [comment]
> and possibly disabling CORS on all routes

Do you mean enabling or disabling? We have CORS on most routes, and we can't disable CORS on them, or else our web apps will be broken (the apps are on a separated origin from the API routes, https://foo-drive.mycozycloud vs https://foo.mycozy.cloud and need CORS).

> CORS are only active on a standard browser anyway and can already be bypassed. The real protection comes from the tokens embeded in the requests — except on the routes related to login/registrations of course. From what I know, it's pretty standard practice to enable CORS on API servers.

It's true, but I don't see what you want here. All the endpoint of cozy-stack protected by a token have CORS enabled. It's just the routes related to login/registrations (`/auth`) that don't have that.

**Comment by y-lohse** [comment]
I meant allowing cross-origin requests on all routes — so yeah it's already the case, my bad. Perefect, let's not change that!

So the only change we need is to also allow cross-origin requests on the `/auth` routes!

**Comment by nono** [comment]
> So the only change we need is to also allow cross-origin requests on the /auth routes!

No! It's a very bad idea from a security point of view to enable CORS for all the `/auth` routes (login and logout in particular). There is the special case of `/auth/register`, where we may do that, but I don't have clear ideas on what are the security implications (and I didn't have time to look at this since the issue is opened).

**Comment by y-lohse** [comment]
Well, we essentially want to do [this](https://docs.cozy.io/en/cozy-stack/auth/#third-party-websites), which isn't possible at the moment. We'd definitely need `/auth/register`, but I think we also need `/auth/access_token` to complete the oauth flow. The other `/auth` routes might not be needed for basic use cases.

> and I didn't have time to look at this since the issue is opened

No problem :)

**Comment by ptbrowne** [comment]
Ping, it is a pain to test the OAuth flow as we need to be on mobile. It would be easier if possible to do this from a browser and would allow to test from TestCafe. Last problem I had trying to open the OAuth page was that the OPTIONS method is not allowed on `/auth/register`. Can we move on this problem ?

**Comment by nono** [comment]
Pong, I'm looking at this.

**Comment by nono** [comment]
I close this issue as WONTFIX.

Currently, we allow OAuth with the dynamic registration protocol extension and with only the authorization code flow. It works for mobile, browser extensions, and websites with a server part. For single-page apps, OAuth has an implicit flow, but it is now deprecated: there are some known attacks, and the people working on OAuth have tried to fix the attacks, but haven't find a good solution do that. It seems to be a really hard problem, and https://tools.ietf.org/id/draft-parecki-oauth-browser-based-apps-02.txt for example says to use a server or not to use OAuth at all.

Github didn't allow the implicit flow for their API, even if it is often asked. A work-around is to use a server component like https://github.com/prose/gatekeeper. Note that it requires a pre-registered client to configure it, but it would be possible to have a similar proxy for Cozy with an additional route to register the client : 

1. The user opens the SPA
2. The user gives its Cozy instance URL
3. The SPA calls the proxy (`POST /client?cozy=...`)
4. The proxy calls the Cozy (`POST /auth/register`)
5. The proxy saves the response from the Cozy (in particular the client_secret)
6. The proxy responds to the SPA with the client_id
7. The SPA redirects the user to the `/auth/authorize?client_id=...&...`
8. The user accepts to give the permission to access the data
9. The user is redirected to the SPA
10. The SPA calls the proxy with the code in the redirected URL (`POST /token?code=...`)
11. The proxy calls the Cozy with the client_secret and the code (`POST /auth/access_token`)
12. The proxy gets the access_token from the response and give it to the SPA
13. The SPA can make calls to the Cozy with this token
14. Profit!
